src/sagemaker_xgboost_container/mms_patch/model_server.py

```diff
@@ -0,0 +1,178 @@
+# language governing permissions and limitations under the License.
+from __future__ import absolute_import
+import os
+import signal
+import subprocess
+import sys
+
+import pkg_resources
+import psutil
+from retrying import retry
+
+import sagemaker_inference
+from sagemaker_inference import default_handler_service, environment, logging, utils
+from sagemaker_inference.environment import code_dir
+
+logger = logging.get_logger()
+
+DEFAULT_HANDLER_SERVICE = default_handler_service.__name__
+MMS_CONFIG_FILE = os.path.join('/etc', 'sagemaker-mms.properties')
+DEFAULT_MMS_CONFIG_FILE = pkg_resources.resource_filename(sagemaker_inference.__name__, '/etc/default-mms.properties')
+DEFAULT_MMS_LOG_FILE = pkg_resources.resource_filename(sagemaker_inference.__name__, '/etc/log4j.properties')
+DEFAULT_MMS_MODEL_DIRECTORY = os.path.join(os.getcwd(), '.sagemaker/mms/models')
+DEFAULT_MMS_MODEL_NAME = 'model'
+
+PYTHON_PATH_ENV = 'PYTHONPATH'
+REQUIREMENTS_PATH = os.path.join(code_dir, "requirements.txt")
+MMS_NAMESPACE = "com.amazonaws.ml.mms.ModelServer"
+
+
+def start_model_server(is_multi_model=False, handler_service=DEFAULT_HANDLER_SERVICE, config_file=None):
+    """Configure and start the model server.
+
+    Args:
+        is_multi_model (bool): Whether to start MxNet Model Server as single model or multi model. If
+        handler_service (str): python path pointing to a module that defines a class with the following:
+            - A ``handle`` method, which is invoked for all incoming inference requests to the model server.
+            - A ``initialize`` method, which is invoked at model server start up for loading the model.
+            Defaults to ``sagemaker_inference.default_handler_service``.
+        config_file (str): path to user defined MMS properties file
+
+    """
+    if not config_file:
+        _create_model_server_config_file()
+        config_file = MMS_CONFIG_FILE
+
+    if os.path.exists(REQUIREMENTS_PATH):
+        _install_requirements()
+
+    _set_python_path()
+
+    mxnet_model_server_cmd = ['mxnet-model-server',
+                              '--start',
+                              '--mms-config', config_file,
+                              '--log-config', DEFAULT_MMS_LOG_FILE,
+                              ]
+
+    if not is_multi_model:
+        _adapt_to_mms_format(handler_service)
+        mxnet_model_server_cmd += ['--model-store', DEFAULT_MMS_MODEL_DIRECTORY]
+
+    logger.info(mxnet_model_server_cmd)
+    subprocess.Popen(mxnet_model_server_cmd)
+
+    mms_process = _retrieve_mms_server_process()
+
+    _add_sigterm_handler(mms_process)
+
+    mms_process.wait()
+
+
+def _adapt_to_mms_format(handler_service):
+    """Archive initial model using MMS handler
+
+    :param handler_service:
+    :return:
+    """
+    if not os.path.exists(DEFAULT_MMS_MODEL_DIRECTORY):
+        os.makedirs(DEFAULT_MMS_MODEL_DIRECTORY)
+
+    model_archiver_cmd = ['model-archiver',
```

Comment:
Why do you need to run model-archiver? I would recommend not using this and instead using ```default_service_handler``` (which is already used in the config.properties). The model-archiver command might add to the initialization latency, as it tries to zip or tar all the artifacts.

Reply:
Right. I actually tried running this without and I was having some trouble last week. I am ok revisiting this, but I kept it for now because eventually I wanted to merge these changes to upstream sagemaker-inference-toolkit which does use this; I'm not sure what I would break if I simply removed it there.

Reply:
Sure. Makes sense.